### ISSUE TITLE
AttributeError: 'BatchNormalizationFunction' object has no attribute 'fixed_mean'

### ISSUE THREAD

**Issue by abcteo** [chainer/chainer#2515]
I have tried the BatchNormalizationFunction with chainer 1.22 and encounter this error
pls help. Thanks

**Comment by delta2323** [comment]
Thank you reporting the issue. Is it possible for you to give us the minimum script so that we can reproduce the error? Thanks.

**Comment by niboshi** [comment]
Reproduced with this code:

```
from chainer.functions.normalization import batch_normalization
import numpy

batch_normalization.BatchNormalizationFunction()(
    numpy.ones((1,1), numpy.float64),
    numpy.array([1.0]),
    numpy.array([1.0]))
```

Result:
```
Traceback (most recent call last):
  File "test_bn.py", line 7, in <module>
    numpy.array([1.0]))
  File "/home/niboshi/w/repos/chainer/chainer/function.py", line 199, in __call__
    outputs = self.forward(in_data)
  File "/home/niboshi/w/repos/chainer/chainer/functions/normalization/batch_normalization.py", line 165, in forward
    mean = self.fixed_mean
AttributeError: 'BatchNormalizationFunction' object has no attribute 'fixed_mean'
```

**Comment by delta2323** [comment]
@niboshi Thank you for your investigation. I will check the code.

**Comment by delta2323** [comment]
It occurs when the length of `inputs` argument of the `forward` method is 3 and `self.train` is `False` when `forward` is called. This function seems to assume this case does not occur.

**Comment by delta2323** [comment]
Proposal for the improvement:

||`len(inputs)==5`|`len(inputs)==3`|
|-------|-------|-------|
|`self.train==True`|||
|`self.train==False`|||

**Comment by niboshi** [comment]
I think we can also raise an error in the top-left case.
It's enforcement of specification, not breaking compatibility.

**Comment by niboshi** [comment]
And in v2, the default `configuration.train` value is `True` in contrary to `BatchNormalizationFunction`'s default `train`  value of `False` in v1. We should inform the user in a strong way of this change, lest the user ignore the warning.

**Comment by delta2323** [comment]
> It's enforcement of specification, not breaking compatibility.

First of all, I think raising an error in these situation is a change of undocumented behavior, rather than enforcement of specification because `F.BatchNormalization` is not documented and therefore, we do not have specification to enforce for this function.

Theoretically speaking, we can change the behavior of `F.BatchNormalization` anyway because it is not documented and therefore, we do not have any compatibility to be kept. In that sense, I agree that raising an error does not break compatibility.

But users' code will be broken if `F.BatchNormalization` is instantiated directly (i.e. not via `F.batch_normalization` or `F.fixed_batch_normalization`) and its `__call__` method is called under these conditions. I am not sure how many users use this function in this way, but I think we should not break compatibility by raising an error unless we have a reason that we strongly want to forbid this usage.

**Comment by delta2323** [comment]
Regarding the update to v2, the behavior does not change if users use `F.BatchNormalization` via `F.batch_normalization` and `F.fixed_batch_normalization`. For `F.batch_normalization`, it sets `train` to `True` in v1, which is the default value of `config.train`. For `F.fixed_batch_normalization`, it sets `train` to `False` in v1 and switches `config.train` to `False` [here](https://github.com/pfnet/chainer/blob/_v2/chainer/functions/normalization/batch_normalization.py#L363) in v2.

**Comment by niboshi** [comment]
Let's separate the cases.

1. Called via `F.batch_normalization` or `F.fixed_batch_normalization`
 -- No problem, both in v1 and v2

2. Directly call `F.BatchNormalizationFunction` with explicit `train` argument
 -- Unexpected keyword argument in v2

3. Directly call `F.BatchNormalizationFunction` without `train` argument
  -- If `len(inputs) == 5`, it will run in `train=False` mode in v1, and `train=True` mode in v2
  -- If `len(inputs) == 3`, it will not run (bug) in v1, and will run in `train=True` mode in v2

Now I'm focusing on `len(inputs) == 5` case in 3 (maybe this is rare, though).
The point is, it will run in different semantics between v1 and v2.
IMO it's better not to allow this code to run, even with warning, because it's almost certain that the user did not expect the new semantics.

The specification is not documented as you said, but it's implicitly assumed in the code. IMO it's OK to make it explicit.

**Comment by delta2323** [comment]
I come to think it is OK to forbid `len(inputs) == 5 && train==True` case because `backward` method already forbids that. Although it is possible that users could employ this function for only forward propagation (e.g. inference mode), I think it is natural that the assumptions for the inputs are consistent between forward and backward.

The benefit of warning solution is that we can use this function in the same way in training and test mode. But I feel transparency and ignorance of the change of semantics is like two sides of a same coin. If we can use some function in the same way in the different context, we cannot notice even if  the context is changed. Considering that we want to avoid implicit changes of semantics, it is better to forbid at the const of transparency.

**Comment by delta2323** [comment]
@niboshi  Let's resume the discussion, shall we?

**Comment by niboshi** [comment]
Does this problem still occur?
Looks like this has been fixed in #3275?

**Comment by delta2323** [comment]
Oh, I missed it. Then I'll close it.
